v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/schema/SessionBasedMapper.java

```diff
@@ -0,0 +1,100 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.spanner.migrations.schema;
+
+import com.google.cloud.teleport.v2.spanner.ddl.Column;
+import com.google.cloud.teleport.v2.spanner.ddl.Ddl;
+import com.google.cloud.teleport.v2.spanner.ddl.Table;
+import com.google.cloud.teleport.v2.spanner.type.Type;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+public class SessionBasedMapper implements ISchemaMapper {
+
+  Ddl ddl;
+
+  Schema schema;
+
+  public SessionBasedMapper(Schema schema, Ddl ddl) {
+    this.schema = schema;
+    this.ddl = ddl;
+  }
+
+  @Override
+  public String getSpannerTableName(String srcTable) throws NoSuchElementException {
+    Map<String, NameAndCols> toSpanner = schema.getToSpanner();
+    if (!toSpanner.containsKey(srcTable)) {
+      throw new NoSuchElementException(String.format("Source table %s not found", srcTable));
+    }
+    return toSpanner.get(srcTable).getName();
+  }
+
+  @Override
+  public String getSpannerColumnName(String srcTable, String srcColumn)
+      throws NoSuchElementException {
+    Map<String, NameAndCols> toSpanner = schema.getToSpanner();
+    if (!toSpanner.containsKey(srcTable)) {
+      throw new NoSuchElementException(String.format("Source table %s not found", srcTable));
+    }
+    Map<String, String> cols = toSpanner.get(srcTable).getCols();
+    if (!cols.containsKey(srcColumn)) {
+      throw new NoSuchElementException(
+          String.format("Source column %s not found for table %s", srcColumn, srcTable));
+    }
+    return cols.get(srcColumn);
+  }
+
+  @Override
+  public String getSourceColumnName(String spannerTable, String spannerColumn)
+      throws NoSuchElementException {
+    Map<String, NameAndCols> toSource = schema.getToSource();
+    if (!toSource.containsKey(spannerTable)) {
+      throw new NoSuchElementException(String.format("Spanner table %s not found", spannerTable));
+    }
+    Map<String, String> cols = toSource.get(spannerTable).getCols();
+    if (!cols.containsKey(spannerColumn)) {
+      throw new NoSuchElementException(
+          String.format("Spanner column %s not found for table %s", spannerColumn, spannerTable));
+    }
+    return cols.get(spannerColumn);
+  }
+
+  @Override
+  public Type getSpannerColumnType(String spannerTable, String spannerColumn)
+      throws NoSuchElementException {
+    Table spTable = ddl.table(spannerTable);
+    if (spTable == null) {
```

Comment:
Can this check be done once in the constructor for all tables that will be migrated?

Reply:
Whenever we run into these situations, it will be only be due to an invalid session file. Because even table names, columns etc will be fetched from the session file itself. We only receive the the src table name in the changeEvent which is the only place where an external input can be incorrect which we should catch. 

However, we have had customers in the past who tried to modify the session file manually (and incorrectly) which caused the template to fail silently, which could've been avoided if we caught mismatches `chanveEvent.srcColumnName` and columns present in the session.

Reply:
What I meant is - we have the Ddl object, we have the session file - this validation can be done once in the lifetime of the pipeline - at the time of constructor of this  class. We do not need to do it for every call to this function. Basically fail fast.

Reply:
SGTM will add a verification to fail fast which will verify the Spanner Table names/col names are consistent with DDL. 
Not adding verification on whether session types  and ddl types are consistent, because if I do, we will not support the use case that users can end up changing types later on which was a common use case as per the last discussion i.e. it is common for users to have different types on Spanner than on session file.